### DISCUSSION TITLE
How do I get the ORM object back from an upsert statement?

### DISCUSSION THREAD

**Discussion by amjith** [sqlalchemy/sqlalchemy#9286] [Usage Questions]
I am using SQLAlchemy 1.4.41 with SQLite (3.33.0). 

```
def add_activity(db: Session, activity: Activity):
    stmt = insert(Activity).values(activity.dict(exclude={"id", "created_on", "updated_on"}))
    # Upsert `activity` if it exists already
    stmt = stmt.on_conflict_do_update(index_elements=["user_id", "strava_id"], set_=activity.dict(exclude={"id", "created_on", "updated_on"}))
    db_activity = db.execute(stmt)
    db.commit()
    return db_activity
```

How do I get the resulting ORM object from this upsert operation?

**Comment by zzzeek** [comment]
for 1.4 there's a recipe to get ORM objects back from upsert at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-postgresql-on-conflict-with-returning-to-return-upserted-orm-objects (it's the same idea for SQLite upsert, MySQL ON DUPLICATE, etc., all the same thing)

strongly recommend upgrade to 2.0 where this feature is included natively at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-upsert-statements

**Reply by CaselIT** [reply]
In any case you will need to update sqlite, since returning is supported only from version 3.35

**Reply by amjith** [reply]
I upgraded to 3.40.1 and I'm still getting a `sqlalchemy.exc.CompileError: RETURNING is not supported by this dialect's statement compiler` error. 

I'm guessing that SQLAlchemy 1.4 doesn't support `RETURNING` clause for the `sqlite` dialect. 

I am using `from sqlalchemy.dialects.sqlite import insert` to implement the recipe that Mike linked from the docs. 

Here's the full code for refrerence: 

```
from sqlalchemy.dialects.sqlite import insert


def add_activity(db: Session, activity: Activity) -> Activity:
    stmt = insert(Activity).values(activity.dict(exclude={"id", "created_on", "updated_on"}))
    stmt = stmt.on_conflict_do_update(index_elements=["user_id", "strava_id"], set_=activity.dict(exclude={"id", "created_on", "updated_on"})).returning(Activity)
    db.execute(stmt)
    db.commit()
    orm_stmt = select(Activity).from_statement(stmt).execution_options(populate_existing=True)
    db_activity = db.execute(orm_stmt).scalar()
    return db_activity
```

If this is not a supported feature that's totally fine, I will look into upgrading SQLAlchemy to 2.0 in the near future

**Reply by zzzeek** [reply]
sorry, yes I forgot that the entire RETURNING feature for SQLite is only supported by SQLAlchemy 2.0.
